test/testers.jl

```diff
@@ -42,62 +42,42 @@ primalapprox(x) = x
         end
     end
 
-    @testset "Inplace accumumulation: first on Array" begin
+    @testset "Inplace accumulation: identity on Array" begin
         @testset "Correct definitions" begin
-            function ChainRulesCore.frule((_, ẋ), ::typeof(first), x::Array)
-                ẏ = InplaceableThunk(
-                    @thunk(first(ẋ)),
-                    ȧ -> ȧ + first(ẋ),  # This won't actually happen inplace
-                )
-                return first(x), ẏ
+            function ChainRulesCore.frule((_, ẋ), ::typeof(identity), x::Array)
+                ẏ = InplaceableThunk(@thunk(ẋ), ȧ -> ȧ .+= ẋ)
+                return identity(x), ẏ
             end
-            function ChainRulesCore.rrule(::typeof(first), x::Array{T}) where T
-                x_dims = size(x)
-                function first_pullback(ȳ)
-                    x̄_ret = InplaceableThunk(
-                        Thunk() do
-                            x̄ = zeros(T, x_dims)
-                            x̄[1]=ȳ
-                            x̄
-                        end,
-                        ā -> (ā[1] += ȳ; ā)
-                    )
+            function ChainRulesCore.rrule(::typeof(identity), x::Array)
+                function identity_pullback(ȳ)
+                    x̄_ret = InplaceableThunk(@thunk(ȳ), ā -> ā .+= ȳ)
                     return (NO_FIELDS, x̄_ret)
                 end
-                return first(x), first_pullback
+                return identity(x), identity_pullback
             end
 
-            frule_test(first, (randn(4), randn(4)))
-            rrule_test(first, randn(), (randn(4), randn(4)))
+            frule_test(identity, (randn(4), randn(4)))
+            rrule_test(identity, randn(4), (randn(4), randn(4)))
         end
 
-        @testset "Incorrect inplace definitions" begin
-            my_first(value) = first(value)  # we are going to define bad rules on this
-            function ChainRulesCore.frule((_, ẋ), ::typeof(my_first), x::Array)
-                ẏ = InplaceableThunk(
-                    @thunk(first(ẋ)),  # correct
-                    ȧ -> ȧ + 1000*first(ẋ),  # incorrect (also not actually inplace)
-                )
-                return first(x), ẏ
+        @testset "Incorrect in-place definitions" begin
+            my_identity(value) = value  # we will define bad rules on this
+            function ChainRulesCore.frule((_, ẋ), ::typeof(my_identity), x::Array)
+                # only the in-place part is incorrect
```

Comment:
I guess the point is that we are checking whether inplace accumulation actually happens?

The test succeeds if the result is unthunked, but not if inplace accumulation is used?

Reply:
correct